src/mappers/mapper34.asm

```diff
@@ -0,0 +1,78 @@
+// Mapper 34: BNROM
+
+scope Mapper34: {
+Init:
+  addi sp, 8
+  sw ra, -8(sp)
+
+// 32K
+  jal TLB.AllocateVaddr
+  lui a0, 0x1'0000 >> 16  // align 64K to leave a 32k guard page unmapped
+
+  ls_gp(sw a0, mapper34_prgrom_vaddr)
+  ls_gp(sb a1, mapper34_prgrom_tlb_index)
+
+// 0x8000-0x1'0000
+  addi t0, a0, -0x8000
+  la_gp(t1, Write)
+  lli t2, 0
+  lli t3, 0x80
+
+-
+  sw t0, cpu_read_map + 0x80 * 4 (t2)
+  sw t1, cpu_write_map + 0x80 * 4 (t2)
+  addi t3, -1
+  bnez t3,-
+  addi t2, 4
+
+// Hard wired CHR mapping 0x0000-0x2000 (8K)
+  ls_gp(lw t0, chrrom_start)
+  lli t1, 8
+  lli t2, 0
+-
+  sw t0, ppu_map (t2)
+  addi t1, -1
+  bnez t1,-
+  addi t2, 4
+
+// Initially map PRG ROM bank 0, Nametable 0
+  j Write_alt
+  lli cpu_t0, 0
+
+Write:
+// cpu_t0: data
+// cpu_t1: address (unused)
+  addi sp, 8
+  sw ra, -8(sp)
+
+Write_alt:
+  // and t0, cpu_t0, 3
+  sll t0, cpu_t0, 15  // choose from 4 32k PRG ROM banks
+
+  ls_gp(lw a1, prgrom_start_phys)
+  ls_gp(lw a0, mapper34_prgrom_vaddr)
+  ls_gp(lwu t1, prgrom_mask)
+  ls_gp(lbu t2, mapper34_prgrom_tlb_index)
+  
+  and t0, t1
+  add a1, t0
+
+  jal TLB.Map32K
+  mtc0 t2, Index
+
+  la_gp(a0, ppu_ram)
```

Comment:
This whole section setting mirroring isn't needed, [rom.asm set up default mirroring](https://github.com/hcs64/neon64v2/blob/6b2c194efa06743d2a722016828afe3bf076c85d/src/rom.asm#L165) which is sufficient for BNROM. You can have the TLB.Map32K call above do the return with something like:
```
  lw ra, -8(sp)
  addi sp, -8
// Tail call
  j TLB.Map32K
  mtc0 t2, Index
```

In fact the `Write` handler doesn't even need to save `ra` to the stack given how simple this is, so you could do the `lw ra, -8(sp)` / `addi sp, -8` before line 38, remove 43-47, and just return through the tail call of `j TLB.Map32K`.

Reply:
This was meant to point at line 63 and onwards.

Reply:
good idea, done and done